realm/realm-library/src/objectServer/java/io/realm/PermissionManager.java

```diff
@@ -1041,6 +1047,49 @@ private ErrorCode combineErrorCodes(Map<String, ObjectServerError> errors) {
 
     }
 
+    // Task responsible for loading the Permissions result and returning it to the user.
+    // The Permission result is not considered available until the query has completed.
+    private class GetOffersAsyncTask extends PermissionManagerTask<RealmResults<Permission>> {
+
+        private final OffersCallback callback;
+        // Prevent permissions from being GC'ed until fully loaded.
+        private RealmResults<PermissionOffer> loadingOffers;
+
+        GetOffersAsyncTask(PermissionManager permissionManager, OffersCallback callback) {
+            super(permissionManager, callback);
+            this.callback = callback;
+        }
+
+        @Override
+        public void run() {
+            if (checkAndReportInvalidState()) { return; }
+            if (offers != null) {
+                notifyCallbackWithSuccess(offers);
+            } else {
+                // We only want offers that have been created.
+                loadingOffers = managementRealm.where(PermissionOffer.class)
+                        .equalTo("statusCode", 0)
+                        .findAllAsync();
+                loadingOffers.addChangeListener(new RealmChangeListener<RealmResults<PermissionOffer>>() {
+                    @Override
+                    public void onChange(RealmResults<PermissionOffer> loadedOffers) {
+                        loadedOffers.removeChangeListener(this);
+                        if (checkAndReportInvalidState()) { return; }
+                        if (offers == null) {
+                            offers = loadedOffers;
+                        }
+                        notifyCallbackWithSuccess(offers);
+                    }
+                });
+            }
+        }
+
+        private void notifyCallbackWithSuccess(RealmResults<PermissionOffer> permissions) {
+            callback.onSuccess(permissions);
```

Comment:
Is this the right order?  What will happen if the callback throws?  If it doesn't create an inconsistent state, in the callback, it might be better to do the cleanup first.  Else maybe at `try {} finally {}`?